Review typography

nicksp opened the pull request:
Closes #42

Unifies and updates typography usage in the UI lib.

nicksp commented:
@bernardodiasc @joshwnj This is ready for review.

bernardodiasc commented on .snapguidist/__snapshots__/Roadmap-1.snap:
```diff
@@ -49,17 +49,17 @@ exports[`Roadmap-1 1`] = `
+          className="AutoUI_Text-26">
           <h2
-            className="AutoUI_Text-39">
+            className="undefined AutoUI_typo-102 AutoUI_typo-53 AutoUI_typo-22">
```
There are many `undefined` classes in the snapshots, this can be caused by some inexistent class.

nicksp replied:
Good catch mate. Let me double-check that and fix.

bernardodiasc commented on src/components/CollapsibleSection.js:
```diff
@@ -80,6 +80,10 @@ const Content = elem.div(cmz(`
     width: calc(100% - 200px)
   }
 
+  & p {
```
Is there a guarantee that all text blocks will be used with `<p>`? I think it should, that's what's expect, but we need make sure of that, if someone includes text body inside only `<div>` the spacing harmony may result differently.

nicksp replied:
Yea, I thought about that. Not sure how to make it work for each case and just for the child, as it's passed by the end user.

bernardodiasc commented on src/components/ApplicantBadge.js:
```diff
@@ -37,11 +39,13 @@ const Info = elem.div(cmz(`
-  family.smallHeading, `
+  typo.smallHeading, `
   letter-spacing: -.025em
   font-size: 1.2em
```
We have font-size and other props overriding the `typo.smallHeading` defaults, is this really desired?

I'm also worried about the usage of `em` and `rem` (relative units) here, since they will cause difference depending on the context. For instance, if this component is included inside another component that set the font-size to something different from the body (I guess is 20px https://github.com/x-team/auto-ui/blob/fa5a878855e247925381ceb5fc671ca0c01b9152/src/styles/typo.js#L142), it might result different from expected. The quick solution would be use `rem` instead, if still wants to use relative units. But I still unsure if these overidings should happen at all.

bernardodiasc commented on .snapguidist/__snapshots__/Roadmap-1.snap:
```diff
@@ -49,17 +49,17 @@ exports[`Roadmap-1 1`] = `
+          className="AutoUI_Text-26">
           <h2
-            className="AutoUI_Text-39">
+            className="undefined AutoUI_typo-102 AutoUI_typo-53 AutoUI_typo-22">
```
There are many `undefined` classes in the snapshots, this can be caused by some inexistent class.

nicksp replied:
Should be fixed now in new spec files.

bernardodiasc commented on src/components/ApplicantBadge.js:
```diff
@@ -37,11 +39,13 @@ const Info = elem.div(cmz(`
-  family.smallHeading, `
+  typo.smallHeading, `
   letter-spacing: -.025em
   font-size: 1.2em
```
We have font-size and other props overriding the `typo.smallHeading` defaults, is this really desired?

I'm also worried about the usage of `em` and `rem` (relative units) here, since they will cause difference depending on the context. For instance, if this component is included inside another component that set the font-size to something different from the body (I guess is 20px https://github.com/x-team/auto-ui/blob/fa5a878855e247925381ceb5fc671ca0c01b9152/src/styles/typo.js#L142), it might result different from expected. The quick solution would be use `rem` instead, if still wants to use relative units. But I still unsure if these overidings should happen at all.

nicksp replied:
Updated.

bernardodiasc commented on src/components/CollapsibleSection.js:
```diff
@@ -80,6 +80,10 @@ const Content = elem.div(cmz(`
     width: calc(100% - 200px)
   }
 
+  & p {
```
Is there a guarantee that all text blocks will be used with `<p>`? I think it should, that's what's expect, but we need make sure of that, if someone includes text body inside only `<div>` the spacing harmony may result differently.

bernardodiasc commented on src/components/Milestones.js:
```diff
@@ -120,17 +120,18 @@ const styles = {
+    `
+      display: block
+      text-transform: uppercase
+      font-size: 12px
```
This is another case with typography props being included together with predefined values, does that overridings are really needed? Isn't something that could be provided right away from typo sets?

bernardodiasc replied:
I double checked `typo.tinyHeading`, it seems that it doesn't provide `font-size` and `color` by default, I guess it's not an override case here. Will leave the comment anyway if we can think in a way to move that to vars, or disregard the comment if it's just fine how it is. :)

nicksp replied:
Moved that over to the typo set. But, given the `Button` component assumes dynamic sizing, still keep font-sizes there to override the one coming from typo set.

nicksp commented:
Just wanted to make a quick note here. Regarding smth. that overrides fonts and other typography related props: If it's meant to be used on admin pages and our Sketch designs don't cover that, we shouldn't worry about that for now.

bernardodiasc commented on src/components/SolutionForm/index.js:
```diff
@@ -46,53 +50,32 @@ const Form = elem.form([
+      resize: none;
+      border: 1px solid ${theme.lineSilver3};
+      border-radius: 4px;
+      font-size: 18px;
```
This font-size here are really needed? I see this class inherits the typo.base, which has font-size = 20px, so I see that this is a variation, but question is: isn't better to avoid these overidings in favor of all typography styles consistently defined in the typo.js?

I understand if isn't the case, just making questions here :)

nicksp replied:
This is intentional. As per Sketch design, the base text (`typo.base`) is `20px`. But form values differ from the regular text by font-size only. I thought instead of specifying each possible typography definition, that may vary from other just slightly, I'd go with overrides. But after another examination of styles, and your note here, I realized, I'll define another typo class that will be composed of the base with some overrides. This way we'll have a single place to maintain typography.

nicksp replied:
Fixed that by defining a new typography set `typo.formText`

bernardodiasc commented on src/components/Button.js:
```diff
@@ -24,28 +24,24 @@ type Props = {
+    color: ${theme.baseBrighter};
+    cursor: pointer;
+    display: inline-block;
+    font-weight: 400;
```
Maybe the typography props here could be provided by typo variables right away.

nicksp replied:
Fixed during revising the `Button` component.

bernardodiasc commented on src/components/Button.js:
```diff
@@ -95,12 +91,7 @@ const sizeStyles = {
-      font-size: 1em;
-    `
-  ),
+  large: cmz(`font-size: 1em`),
```
Be careful with relative units, if this `em` goes in a non-root element, it possibly can result the expected size on first sight. But if the `em` is being used on the root element of the component, the results can be unpredictable, because will vary accordingly the font-size of the first parent that has font-size defined, that is unknown from this component perspective.

The easy solution is to set `rem` in the root element of the component, that way the unpredictability is narrowed down to the `body` font-size. Since relative units can cause a lot trouble if not using very consciously, I'd recommend to stick to absolute units to avoid that trouble.

nicksp replied:
Yea, fixed that to use absolute units as part of revising `Button` component. Thanks for spotting 👍

bernardodiasc commented:
@nicksp one extra suggestion. You could create a component similar to `ColorPalette` for Typography variables. It's quite simple, just get the variables and loop through including styles in blocks of texts `The quick brown fox jumps over the lazy dog`. This will provide handy visualisation of all the typography sets available. :)

nicksp commented:
@bernardodiasc Created a `Typography` visualisation component, thanks for suggestion 😛 It'll be extended automatically as we add more typography sets :)

![localhost_6060_](https://user-images.githubusercontent.com/579331/33915669-a4ffd106-dfb5-11e7-9db8-ff647f05666e.png)

bernardodiasc commented:
That's awesome! :D

nicksp commented:
@bernardodiasc I revised the complete typography sets and updated accordingly :-) Also I fixed the issue with `undefined`. The ones you still can see come from the old snaps. I guess they don't get updated. Latest snaps don't contain `undefined` as part of class name.

If you got some time, could you please re-review that, so this PR doesn't block others? Just want to get this out the door asap 😝

bernardodiasc commented on src/components/Button.js:
```diff
@@ -91,26 +84,11 @@ const colorStyles = {
-  normal: cmz(`
-    font-size: .75em;
-  `),
+  small: cmz(`font-size: 10px !important`),
```
I imagine why `!important` was required here. It's some kind of trouble I had a lot with CSS Modules as well. We don't have exact control of the order that classes are printed in the output and having low specificity the latest declarations overrides the first ones.

I think there is no much do do about this to avoid the `!important` tag, so I guess that's fine.

nicksp replied:
Right, `!important` here is to ensure exact font sizes are applied over the base ones. I tried getting control over the order of classes, but couldn't find a better way to do that. Will keep that as is for now.

bernardodiasc commented on src/components/CollapsibleSection.js:
```diff
@@ -80,6 +80,10 @@ const Content = elem.div(cmz(`
     width: calc(100% - 200px)
   }
 
+  & p {
```
Yeah, I understand what you said, it's kinda fine this way. Let's keep an eye in how the CollapsibleSection are used, if someone includes text blocks without `<p>` that results visually wrong, then we know where the problem is. :)

nicksp replied:
indeed, let's keep an eye on it :-)
